docs/proposal/kmesh_support_localityLB.md

```diff
@@ -0,0 +1,192 @@
+---
+title: Locality Load Balancing for Kmesh
+authors:
+- "@derekwin" # Authors' GitHub accounts here.
+reviewers:
+- "@kwb0523"
+- "@hzxuzhonghu"
+approvers:
+- "@robot"
+- TBD
+
+creation-date: 2024-06-07
+
+---
+
+## Locality Load Balancing for Kmesh
+
+### Summary
+
+Add Locality Load Balancing to Kmesh workload mode.
+
+### Motivation
+
+Currently, Kmesh does not support locality topology-aware load balancing. Locality Load Balancing optimizes performance and reliability in distributed systems by directing traffic to the nearest service instances. This approach reduces latency, enhances availability, and lowers costs associated with cross-region data transfers. It also ensures compliance with data sovereignty regulations and improves the overall user experience by providing faster and more reliable service responses.
+
+### Goals
+
+The purpose of this proposal is to add locality-based topology-aware load balancing capabilities to kmesh workload mode, corresponding to locality load balancing in Ztunnel.
+
+### Proposal
+
+1. prioritize add locality load balancing capabilities in the workload mode.
+
+2. locality load balancing mode: locality failover.
+
+#### locality failover
+
+1. configure
+
+Configure the load balance type (failover or strict) and scope (REGION, ZONE, etc.) in the service. This corresponds to the message LoadBalancing in workload.proto.
+```
+message LoadBalancing {
+  enum Scope {
+    UNSPECIFIED_SCOPE = 0;
+    REGION = 1;
+    ZONE = 2;
+    SUBZONE = 3;
+    NODE = 4;
+    CLUSTER = 5;
+    NETWORK = 6;
+  }
+  enum Mode {
+    UNSPECIFIED_MODE = 0;
+    STRICT = 1;
+    FAILOVER = 2;
+  }
+  repeated Scope routing_preference = 1;
+  Mode mode = 2;
+}
+```
+https://pkg.go.dev/istio.io/istio/pkg/workloadapi#LoadBalancing_Scope
+
+2. calculate locality match rank
+
+Calculate the rank based on the locality of the src and the locality of all target endpoints in the service according to the configuration in loadbalance. This rank indicates the priority of endpoint selection.
+- The target endpoints are all healthy endpoints.
+
+3. choose endpoint
+
+Randomly select one endpoint from the group with the highest rank as the service backend.
```

Comment:
```suggestion
Randomly select one endpoint from the group with the highest priority
```

And add more comments what we do if all the endpoints of high priority is unhealthy

Reply:
And  for the strict mode, how would you select the endpoint, i would like to see that